### PR TITLE
table io: implement `trace_evaluation`

### PR DESCRIPTION
**Pull request by tcoratger**

### CODE REVIEW

File: crates/brainfuck_prover/src/components/io/table.rs [hunk 1]
```diff
@@ -77,6 +85,51 @@ impl<const N: u32> IOTable<N> {
...
+    ///
+    /// # Errors
+    /// Returns [`TraceError::EmptyTrace`] if the table is empty.
+    pub fn trace_evaluation(&self) -> Result<(TraceEval, Claim), TraceError> {
+        let n_rows = self.table.len() as u32;
+
+        // If the table is empty, return an error.
+        if n_rows == 0 {
+            return Err(TraceError::EmptyTrace);
+        }
```
Comment (zmalatrax) [inline crates/brainfuck_prover/src/components/io/table.rs:107]: There could have programs which don't do any input/output so the I/O table might be empty

Comment (zmalatrax) [reply crates/brainfuck_prover/src/components/io/table.rs:107]: Yeah, I'm not sure on how to handle this case, maybe early return with a default empty TraceEval and a Claim with log_size to 0, wdyt?

Comment (tcoratger) [reply crates/brainfuck_prover/src/components/io/table.rs]: Yes seems logical to me like this and we can treat this trace in a particular way afterwards if we need for this specific case. Just modified it

File: crates/brainfuck_prover/src/components/io/table.rs [hunk 2]
```diff
@@ -77,6 +85,55 @@ impl<const N: u32> IOTable<N> {
...
+    /// The [`IOTable`] is transformed from an array of rows (one element = one step
+    /// of all registers) to an array of columns (one element = all steps of one register).
+    /// It is then evaluated on the circle domain.
+    ///
+    /// # Returns
+    /// A tuple containing the evaluated trace and claim for STARK proof.
+    ///
+    /// # Errors
+    /// Returns [`TraceError::EmptyTrace`] if the table is empty.
+    pub fn trace_evaluation(&self) -> Result<(TraceEval, Claim), TraceError> {
```
Comment (zmalatrax) [inline crates/brainfuck_prover/src/components/io/table.rs:101]: As we're not throwing any Error anymore, we might not be obliged to have a `Result<>` output 🤔

File: crates/brainfuck_prover/src/components/io/table.rs [hunk 3]
```diff
@@ -77,6 +85,55 @@ impl<const N: u32> IOTable<N> {
...
+    ///
+    /// The [`IOTable`] is transformed from an array of rows (one element = one step
+    /// of all registers) to an array of columns (one element = all steps of one register).
+    /// It is then evaluated on the circle domain.
+    ///
+    /// # Returns
+    /// A tuple containing the evaluated trace and claim for STARK proof.
+    ///
+    /// # Errors
+    /// Returns [`TraceError::EmptyTrace`] if the table is empty.
```
Comment (zmalatrax) [inline crates/brainfuck_prover/src/components/io/table.rs:100]: Update the header doc

File: crates/brainfuck_prover/src/components/io/table.rs [hunk 4]
```diff
@@ -77,6 +85,55 @@ impl<const N: u32> IOTable<N> {
...
+    pub fn trace_evaluation(&self) -> Result<(TraceEval, Claim), TraceError> {
+        let n_rows = self.table.len() as u32;
+
+        // It is possible that the table is empty because the program has no input or output.
+        //
+        // In this case, we return:
+        // - An empty trace.
+        // - A claim with a log size of 0.
```
Comment (zmalatrax) [inline crates/brainfuck_prover/src/components/io/table.rs:108]: This part can be refactored into the `# Returns` part of the header doc

File: crates/brainfuck_prover/src/components/io/table.rs (hunk 3 above)
Comment (tcoratger) [reply crates/brainfuck_prover/src/components/io/table.rs:100]: true, fixed

File: crates/brainfuck_prover/src/components/io/table.rs (hunk 2 above)
Comment (tcoratger) [reply crates/brainfuck_prover/src/components/io/table.rs:101]: fixed

File: crates/brainfuck_prover/src/components/io/table.rs (hunk 4 above)
Comment (tcoratger) [reply crates/brainfuck_prover/src/components/io/table.rs:108]: fixed

**Review by zmalatrax — APPROVED** [review]

**Pull request merged**